sdks/java/core/src/main/java/com/google/cloud/dataflow/sdk/io/CountingInput.java

```diff
@@ -157,7 +160,19 @@ public UnboundedCountingInput withTimestampFn(SerializableFunction<Long, Instant
     public UnboundedCountingInput withMaxNumRecords(long maxRecords) {
       checkArgument(
           maxRecords > 0, "MaxRecords must be a positive (nonzero) value. Got %s", maxRecords);
-      return new UnboundedCountingInput(timestampFn, Optional.of(maxRecords), maxReadTime);
+      return new UnboundedCountingInput(timestampFn, period, Optional.of(maxRecords), maxReadTime);
+    }
+
+    /**
+     * Returns an {@link UnboundedCountingInput} like this one, but with output production limited
+     * to an aggregate rate of no more than one element per period.
+     *
+     * <p>Note that this period is taken in aggregate across all instances of the
+     * {@link PTransform}, which may cause elements to be produced in bunches.
+     */
+    public UnboundedCountingInput withRate(long numElements, Duration periodLength) {
+      Duration newPeriod = periodLength.dividedBy(numElements);
```

Comment:
Yep. I've changed this to produce batches of elements (of size numElements) every period, in aggregate. This does mean that we get (potentially large) batches of elements every period, but given that this is an aggregate rate I find that acceptable.